Time shortcode refactor and QUnit set up

claudiulodro opened the pull request:
I've refactored the time shortcode module to make it more robust and unit-testable, as well as unit testing the module.

I've also set up a basic QUnit implementation to enable testing of the time shortcode module javascript and other javascript from the plugin.

mdawaffe commented on modules/time-shortcode/load.php:
```diff
@@ -10,114 +10,150 @@
+
+		$html_entity_regex = '/&[^\s]*;/';
+		$date_string = preg_replace( $html_entity_regex, ' ', $date_string );
+		$date_string = str_ireplace( "U+00A0", ' ', $date_string );
```
There's no need to handle the string "U+00A0". The original problem was with a no-break space: Unicode codepoint U+00A0. In UTF-8, that's the byte sequence "\xC2\xA0".

mdawaffe commented on modules/time-shortcode/load.php:
```diff
@@ -10,114 +10,150 @@
 
-	// only process the time shortcode
-	$shortcode_tags = array( 'time' => 'o2_time_shortcode' );
+					var time = $this.data( 'time' );
```
I found this surprising that `jQuery.data()` returns a Number (rather than a String) here. Can we make that clear here? A code comment? Sending it through `Number()`? Something else?

mdawaffe commented on tests/phpunit/modules/test-time-shortcode.php:
```diff
@@ -0,0 +1,205 @@
+		$time_string = 'now';
+		$epoch_now = date( 'U' );
+
+		$parsed_time = o2_Time_Shortcode::parse_time( $time_string );
```
It looks like there's a chance for this test to fail (if the second changes between the `date()` call here and the `strtotime()` call in `::parse_time()`. It probably doesn't matter when humans run the tests (a rerun almost certainly won't trigger it), but robots might get mad. Thoughts?

mdawaffe commented on tests/phpunit/modules/test-time-shortcode.php:
```diff
@@ -0,0 +1,205 @@
+
+	function test_parse_time_invalid() {
+
+		$time_string = '12 Bananuary 2015';
```
Bananuary!

claudiulodro commented:
Great feedback! I've made the changes you requested in the following ways:

1. In the date parser, I run the input through htmlentities then strip out all html entities. This sanitizes unwanted unicode and html entity content at the same time. I've modified the tests to prove this works.

2. I'm now running the data field through Number.parseInt and have added a comment to clarify that part.

3. I added a 1-second max threshold for the "Now" test and another test with a fixed timestamp for better coverage.

I've also gone through and made sure all the coding style matches the guidelines in the WP PHP Coding Standards.

Please let me know if you notice anything or have any problems.

claudiulodro commented:
This pull request now also includes the functionality requested in https://github.com/Automattic/o2/issues/113.

mdawaffe commented:
All looks good to me. Normally, I'd say we should have two separate PRs for these two separate issues, but I think it's OK in this case.

mdawaffe commented:
Two questions about `Number.parseInt()`:
1. Does it exist in older browsers (and do we care)?
2. Does it need the radix parameter set (`Number.parseInt( foo, 10 )`)?

claudiulodro commented:
Good questions. 

Upon further research, it looks like for best compatibility I would want to use the global `parseInt()` instead of `Number.parseInt()`. They are the exact same thing, but `parseInt()` has wider support (compare browser support at [parseInt](https://msdn.microsoft.com/en-us/library/x53yedee%28v=vs.94%29.aspx) vs [Number.parseInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/parseInt)). 

In modern browsers the radix parameter is completely optional. In older browsers, it was recommended to have the radix parameter because the default would parse any numbers beginning with "0" as octal, but current implementations won't do that. It is still considered best-practices to include the radix in the function call.

I will modify that right now to work with the higher-compatibility `parseInt()` using an explicit radix.

claudiulodro commented:
@pento, this pull request has been thoroughly vetted and should be good to go whenever you get a chance. No real hurry; you won't have any merge conflicts unless you start modifying the time shortcode module.  :-)
